File: 155.min-stack.cpp
/*
 * @lc app=leetcode id=155 lang=cpp
 *
 * [155] Min Stack
 */

#include <bits/stdc++.h>

#include <cassert>

using namespace std;

// @lc code=start
class MinStack {
 public:
  /** initialize your data structure here. */
  MinStack() {}

  void push(int x) {
    stack_.push_back(x);
    for (auto it = sorted_.begin(); it != sorted_.end(); ++it) {
      if (x < *it) {
        it = sorted_.insert(it, x);
        return;
      }
    }
    sorted_.push_back(x);
  }

  void pop() {
    int last = *(stack_.end() - 1);
    stack_.pop_back();
    for (auto it = sorted_.begin(); it != sorted_.end(); ++it) {
      if (last == *it) {
        sorted_.erase(it);
        return;
      }
    }
  }

  int top() { return *(stack_.end() - 1); }

  int getMin() { return sorted_[0]; }

 private:
  std::vector<int> stack_;
  std::vector<int> sorted_;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
// @lc code=end

int main() {
  {
    MinStack s;
    s.push(-2);
    assert(s.top() == -2);
    assert(s.getMin() == -2);
    s.push(-3);
    assert(s.top() == -3);
    assert(s.getMin() == -3);
    s.push(0);
    assert(s.top() == 0);
    assert(s.getMin() == -3);
    s.pop();
    assert(s.top() == -3);
    assert(s.getMin() == -3);
    s.pop();
    assert(s.top() == -2);
    assert(s.getMin() == -2);
    s.pop();
  }
  return 0;
}

File: 155.min-stack.rs
/*
 * @lc app=leetcode id=155 lang=rust
 *
 * [155] Min Stack
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[macro_export]
macro_rules! list_node {
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let mut node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
struct MinStack {
    stack: Vec<i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MinStack {
    fn new() -> Self {
        MinStack { stack: vec![] }
    }

    fn push(&mut self, val: i32) {
        self.stack.push(val)
    }

    fn pop(&mut self) {
        self.stack.pop();
    }

    fn top(&self) -> i32 {
        self.stack[self.stack.len() - 1]
    }

    fn get_min(&self) -> i32 {
        let mut min_val = std::i32::MAX;
        for num in &self.stack {
            min_val = std::cmp::min(min_val, *num);
        }
        min_val
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * let obj = MinStack::new();
 * obj.push(val);
 * obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: i32 = obj.get_min();
 */
// @lc code=end

#[test]
fn test1() {
    let mut stack = MinStack::new();
    stack.push(-2);
    stack.push(0);
    stack.push(-3);
    assert_eq!(stack.get_min(), -3);
    stack.pop();
    assert_eq!(stack.top(), 0);
    assert_eq!(stack.get_min(), -2);
}
